lightning-invoice/src/payment.rs

```diff
@@ -466,16 +544,36 @@ where
 		}.map_err(|e| PaymentError::Sending(e))
 	}
 
+	// Takes in a path to have its information stored in `payment_cache`. This is done for paths
+	// that are pending retry.
+	fn process_path_inflight_htlcs(&self, payment_hash: PaymentHash, path: &Vec<RouteHop>) {
+		self.payment_cache.lock().unwrap().entry(payment_hash)
+			.or_insert_with(|| PaymentInfo::new())
+			.paths.push(path.clone());
```

Comment:
hm, currently the `path` that gets passed in here in a reference, but `push` is expecting a value, that's why I  `clone`-d. is that not correct?

Reply:
You can have `process_path_inflight_htlcs` take the path by value instead of by reference to avoid the clone. This is because at call sites you have ownership of the `Route`. See how @TheBlueMatt uses `drain` to take ownership during iteration and `zip` to iterate over both paths and results in parallel here: https://github.com/lightningdevkit/rust-lightning/pull/1643#discussion_r956200113

Reply:
Though maybe a little cleaner with `into_iter` instead of `drain`:

```diff
diff --git a/lightning-invoice/src/payment.rs b/lightning-invoice/src/payment.rs
index 8497ca8c..4f76c709 100644
--- a/lightning-invoice/src/payment.rs
+++ b/lightning-invoice/src/payment.rs
@@ -493,7 +493,7 @@ where
                match send_payment(&route) {
                        Ok(payment_id) => {
                                for path in route.paths {
-                                       self.process_path_inflight_htlcs(payment_hash, &path);
+                                       self.process_path_inflight_htlcs(payment_hash, path);
                                }
                                Ok(payment_id)
                        },
@@ -515,10 +515,10 @@ where
                                        // If a `PartialFailure` event returns a result that is an `Ok()`, it means that
                                        // part of our payment is retried. When we receive `MonitorUpdateFailed`, it
                                        // means that we are still waiting for our channel monitor update to be completed.
-                                       for (idx, result) in results.iter().enumerate() {
+                                       for (path, result) in route.paths.into_iter().zip(results.iter()) {
                                                match result {
                                                        Ok(_) | Err(APIError::MonitorUpdateFailed) => {
-                                                               self.process_path_inflight_htlcs(payment_hash, &route.paths[idx]);
+                                                               self.process_path_inflight_htlcs(payment_hash, path);
                                                        },
                                                        _ => {},
                                                }
@@ -545,7 +545,7 @@ where
 
        // Takes in a path to have its information stored in `payment_cache`. This is done for paths
        // that are pending retry.
-       fn process_path_inflight_htlcs(&self, payment_hash: PaymentHash, path: &Vec<RouteHop>) {
+       fn process_path_inflight_htlcs(&self, payment_hash: PaymentHash, path: Vec<RouteHop>) {
                self.payment_cache.lock().unwrap().entry(payment_hash)
                        .or_insert_with(|| PaymentInfo::new())
                        .paths.push(path.clone());
@@ -603,7 +603,7 @@ where
                match self.payer.retry_payment(&route.as_ref().unwrap(), payment_id) {
                        Ok(()) => {
                                for path in route.unwrap().paths {
-                                       self.process_path_inflight_htlcs(payment_hash, &path);
+                                       self.process_path_inflight_htlcs(payment_hash, path);
                                }
                                Ok(())
                        },
@@ -619,10 +619,10 @@ where
                                // If a `PartialFailure` error contains a result that is an `Ok()`, it means that
                                // part of our payment is retried. When we receive `MonitorUpdateFailed`, it
                                // means that we are still waiting for our channel monitor update to complete.
-                               for (idx, result) in results.iter().enumerate() {
+                               for (path, result) in route.unwrap().paths.into_iter().zip(results.iter()) {
                                        match result {
                                                Ok(_) | Err(APIError::MonitorUpdateFailed) => {
-                                                       self.process_path_inflight_htlcs(payment_hash, &route.as_ref().unwrap().paths[idx]);
+                                                       self.process_path_inflight_htlcs(payment_hash, path);
                                                },
                                                _ => {},
                                        }

```

For reference on the different type of iterators and how rust may infer which one is used:

https://doc.rust-lang.org/std/iter/index.html#the-three-forms-of-iteration
https://doc.rust-lang.org/std/iter/index.html#for-loops-and-intoiterator
https://doc.rust-lang.org/std/iter/index.html#iterating-by-reference

Reply:
ah, I see. thanks for taking the time to put together the diffs and explaining how ownership works. I definitely need to read up more on it, lol